lib/Runtime/Library/JSON.cpp

```diff
@@ -424,85 +427,114 @@ namespace JSON
         return StrHelper(key, value, holder);
     }
 
-    Js::Var StringifySession::Str(Js::JavascriptString* key, Js::PropertyId keyId, Js::Var holder)
+    Js::Var StringifySession::Str(Js::JavascriptString* key, Js::PropertyId keyId, Js::Var holder, Js::Var value)
     {
-        Js::Var value = nullptr;
         // We should look only into object's own properties here. When an object is serialized, only the own properties are considered,
         // the prototype chain is not considered. However, the property names can be selected via an array replacer. In this case
         // ES5 spec doesn't say the property has to own property or even to be enumerable. So, properties from the prototype, or non enum properties,
         // can end up being serialized. Well, that is the ES5 spec word.
         //if(!Js::RecyclableObject::FromVar(holder)->GetType()->GetProperty(holder, keyId, &value))
 
-        if (VirtualTableInfo<Js::PropertyString>::HasVirtualTable(key))
+        if (value == nullptr)
         {
-            PropertyValueInfo info;
-            Js::PropertyString* propertyString = (Js::PropertyString*)key;
-            PropertyValueInfo::SetCacheInfo(&info, propertyString, propertyString->GetLdElemInlineCache(), false);
-            CacheOperators::TryGetProperty<true, false, true, false, true, false, false, true, false>(holder, false, Js::RecyclableObject::FromVar(holder), keyId, &value, scriptContext, nullptr, &info);
+            if (VirtualTableInfo<Js::PropertyString>::HasVirtualTable(key))
+            {
+                PropertyValueInfo info;
+                Js::PropertyString* propertyString = (Js::PropertyString*)key;
+                PropertyValueInfo::SetCacheInfo(&info, propertyString, propertyString->GetLdElemInlineCache(), false);
+                CacheOperators::TryGetProperty<true, false, true, false, true, false, false, true, false>(holder, false, Js::RecyclableObject::FromVar(holder), keyId, &value, scriptContext, nullptr, &info);
+            }
+
+            if (value == nullptr)
+            {
+                if (!Js::JavascriptOperators::GetProperty(Js::RecyclableObject::FromVar(holder), keyId, &value, scriptContext))
+                {
+                    return scriptContext->GetLibrary()->GetUndefined();
+                }
+            }
         }
+        return StrHelper(key, value, holder);
+    }
 
-        if (value == nullptr)
+    inline bool Get_ToJSON(ScriptContext* scriptContext, Js::JavascriptString* key, Js::Var* value, Js::TypeId typeId)
+    {
+        Js::Var toJSON = nullptr;
+        Js::RecyclableObject* object = Js::RecyclableObject::FromVar(*value);
+        while (typeId != Js::TypeIds_Null)
```

Comment:
Instead of doing the prototype walk yourself, why not just call Js::JavascriptOperators::GetProperty here? LTO should take care of optimizing this here, and if the concern is GetProperty does more than we want it to (eg. update the property caches), consider adding a new template parameter to GetProperty_Internal to not update the caches. It's probably better to have property getting centralized to avoid edge cases (eg. what happens if TypeFlagMask_SkipsPrototype is set etc)

Reply:
Well,  I consider this more than a loop hanging around. The logic here is defined under`Get_ToJSON`. It is probably the second hottest place in JSON.stringify and specifics are clear enough. `GetPropertyQuery` is a centralized interface no?

Reply:
What about dealing with `TypeFlagMask_SkipsPrototype`? What is the expected behavior with Proxies?

Reply:
Not sure what is the problem there? Can you please elaborate a case where it fails?

Reply:
In JavascriptOperators::GetProperty_Internal, they explicit break out of the loop if object->SkipsPrototype() is true. I don't quite understand when that flag is used but it looks like its in ChakraFull, so the DOMs type system relies on it. I don't have a ready case that fails but my point is that these are things that JavascriptOperators::GetProperty_Internal that takes into account that you'll need to keep in mind if you're going to reimplement it here, which is why centralizing on JavascriptOperators::GetProperty_Internal is not a bad idea.

Re proxies, again, don't have a failing case- it was more a question of what is the expected behavior.

Reply:
@digitalinfinity this particular `Get_ToJSON`, we know it's an object. So, no need for redundant is number, is that or this check. Also we skip cache as well as other centralized stuff we must check for GetProperty_Internal. 

Lets say we ended up with version of `GetProperty_Internal` that we don't do any of these given above. It could be hard to decide on any place that whether we should use that version ever or not.

I didn't want to rule caching out straight and  implement GetProperty_InternalWithoutCache because at this moment I'm not sure where that cache is beneficial and where it's not. If you prefer we might also add a `TODO` note here but I'm also not sure what that could be.

Reply:
That logic makes sense for number etc but the SkipsPrototype case I mentioned is an edge specific object case- should that be handled here or not?

And I'd still love to understand what the expected behavior here should be for proxies (not saying there's an issue, just want to understand if proxy traps could be invoked as a part of this, in which case there could be other issues).

Reply:
> That logic makes sense for number etc but the SkipsPrototype case I mentioned is an edge specific object case- should that be handled here or not?

I had went through `TypeFlagMask_SkipsPrototype` and couldn't see any reason why we check it for `toJSON`. That object doesn't have `toJSON` in it's body. Well, IMHO we should disable that check for ChakraCore for good. No?

About Proxy; I don't see any reason (on code) why this wouldn't work. There is no proxy trap problem. If the check for `HasSpecialType` returns true and type was proxy, well we end up doing the same call we do here. However, if the type is `null`? or other specialized type (which there is no such a thing exist), well they are not the problem of `Get_toJSON` anyways.